Handle ICMP frag in OVN for packet sizes greater than pod MTU

zshi-redhat opened the pull request:
This PR reverts 7e832266167b518fdb9f6eba63623e7753e753f5 and 78cf7be68841456ac427d0e709280e4cd7e3883b commits,
and fixes the HWOL issue: https://bugzilla.redhat.com/show_bug.cgi?id=2000988
    
Note: it removes the --disable-pkt-mtu-check flag

Signed-off-by: Zenghui Shi <zshi@redhat.com>

zshi-redhat commented:
Fixed the error caused by reverts.

zshi-redhat commented:
/cc @dcbw  @trozet

zshi-redhat commented:
/cc @adrianchiris @girishmg  This fixes the east-west HWOL issue that check_pkt_larger is not offloadable yet.

dcbw commented on dist/images/run-smart-nic-ovnkube-node.sh:
```diff
@@ -10,5 +10,4 @@ docker run --pid host --network host --user=0 --name ovn-node -dit --cap-add=NET
   -e OVN_KUBERNETES_NAMESPACE=ovn-kubernetes -e OVN_SSL_ENABLE=no -e OVNKUBE_NODE_MODE="smart-nic" \
   -e OVN_ENCAP_IP=$SMART_NIC_IP -e K8S_TOKEN=$K8S_TOKEN -e K8S_CACERT=$K8S_CACERT \
   -e OVN_GATEWAY_OPTS="$OVN_GATEWAY_OPTS" -e OVNKUBE_NODE_MGMT_PORT_NETDEV="$OVNKUBE_NODE_MGMT_PORT_NETDEV" \
-  -e OVN_DISABLE_PKT_MTU_CHECK=true \
```
Should this be part of the previous "Handling packet sizes greater than pod MTU" patch?

dcbw replied:
Or did I miss another commit we should revert?

zshi-redhat replied:
This file was added in one of the SmartNIC commits after the OVN_DISABLE_PKT_MTU_CHECK was introduced.
I thought it would be simple to remove this line manually than reverting the SmartNIC one (which results in large conflicts since there is a series of SmartNIC commits).

dcbw replied:
Ok thanks for the explanation. I agree.

dcbw commented on docs/kind.md:
```diff
@@ -84,7 +84,6 @@ configuration options when deploying. Use `./kind.sh -h` to see the latest optio
 usage: kind.sh [[[-cf |--config-file <file>] [-kt|keep-taint] [-ha|--ha-enabled]
                  [-ho |--hybrid-enabled] [-ii|--install-ingress] [-n4|--no-ipv4]
                  [-i6 |--ipv6] [-wk|--num-workers <num>] [-ds|--disable-snat-multiple-gws]
-                 [-dp |--disable-pkt-mtu-check]
```
Should this be part of the previous "Handling packet sizes greater than pod MTU" patch?

zshi-redhat replied:
This change was introduced in a separate commit (9137c9a31196373e2fdadaa2917383d8b3f51828) which aims to improve kind documentation, not mtu specific commit.

dcbw commented on docs/kind.md:
```diff
@@ -102,7 +101,6 @@ usage: kind.sh [[[-cf |--config-file <file>] [-kt|keep-taint] [-ha|--ha-enabled]
 -ha  | --ha-enabled                Enable high availability. DEFAULT: HA Disabled.
 -ho  | --hybrid-enabled            Enable hybrid overlay. DEFAULT: Disabled.
 -ds  | --disable-snat-multiple-gws Disable SNAT for multiple gws. DEFAULT: Disabled.
--dp  | --disable-pkt-mtu-check     Disable checking packet size greater than MTU. Default: Disabled
```
Should this be part of the previous "Handling packet sizes greater than pod MTU" patch?

zshi-redhat replied:
This change was introduced in a separate commit (9137c9a) which aims to improve kind documentation, not mtu specific commit.

dcbw commented:
Hmm, still more failures than I'd expect. Are they flakes or legitimate issues with the PR or OVN's check_pkt_larger implementation?

zshi-redhat commented:
> Hmm, still more failures than I'd expect. Are they flakes or legitimate issues with the PR or OVN's check_pkt_larger implementation?

some are flakes, but others may be legitimate issues, like:
```
[Fail] [sig-network] Networking Granular Checks: Services [It] should function for service endpoints using hostNetwork 
/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
```

zshi-redhat commented:
> > Hmm, still more failures than I'd expect. Are they flakes or legitimate issues with the PR or OVN's check_pkt_larger implementation?
> 
> some are flakes, but others may be legitimate issues, like:
> 
> ```
> [Fail] [sig-network] Networking Granular Checks: Services [It] should function for service endpoints using hostNetwork 
> /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
> ```
> 
> I will take a look.

@dcbw The failed case seems due to the udp server used in the k8s e2e testpod doesn't handle the icmp frag properly.

sender  : ovn pod (node 1) -> ovn -> node-port service -> host pod (node 2)
receiver: host pod (node 2) ->  ovn -> ovn pod (node 1)

The failed flow is on the reply path:
1) sender sends the pkt with size 1900 on node 1
2) it reaches the receiver on node 2
3) receiver on node 2 echo back the pkt without mutating it
4) reply pkt from receiver hit ovn gateway port rtoj- where gateway_mtu option is set to 1400
5) icmp frag message is generated by ovn and sent back to receiver
6) icmp message observed on node 2 physical port where receiver (host pod) is located
7) no fragmented pkt be sent back from receiver.

Compared with not using gateway_mtu option in rtoj- port, the reply pkt is forwarded to kernel which does the pkt fragment (I guess?), so it doesn't require the udp server in testpod to handle icmp frag.

zshi-redhat commented:
> > > Hmm, still more failures than I'd expect. Are they flakes or legitimate issues with the PR or OVN's check_pkt_larger implementation?
> > 
> > 
> > some are flakes, but others may be legitimate issues, like:
> > ```
> > [Fail] [sig-network] Networking Granular Checks: Services [It] should function for service endpoints using hostNetwork 
> > /go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I will take a look.
> 
> @dcbw The failed case seems due to the udp server used in the k8s e2e testpod doesn't handle the icmp frag properly.
> 
> sender : ovn pod (node 1) -> ovn -> node-port service -> host pod (node 2)
> receiver: host pod (node 2) -> ovn -> ovn pod (node 1)
> 
> The failed flow is on the reply path:
> 
> 1. sender sends the pkt with size 1900 on node 1
> 2. it reaches the receiver on node 2
> 3. receiver on node 2 echo back the pkt without mutating it
> 4. reply pkt from receiver hit ovn gateway port rtoj- where gateway_mtu option is set to 1400
> 5. icmp frag message is generated by ovn and sent back to receiver
> 6. icmp message observed on node 2 physical port where receiver (host pod) is located
> 7. no fragmented pkt be sent back from receiver.
> 
> Compared with not using gateway_mtu option in rtoj- port, the reply pkt is forwarded to kernel which does the pkt fragment (I guess?), so it doesn't require the udp server in testpod to handle icmp frag.

@trozet When pkt larger than mtu hit check_pkt_larger in the physical bridge br-ex and be forwarded to host kernel, how does it then get to the ovn pod?

trozet commented:
dont we need an OVN bump to pick up some OVN fix that allows handling for this case?

trozet commented:
"@trozet When pkt larger than mtu hit check_pkt_larger in the physical bridge br-ex and be forwarded to host kernel, how does it then get to the ovn pod?"

@zshi-redhat if I understand your question/scenario correctly...the kernel will reply with ICMP too large and send it back into br-ex bridge, then out the physical nic, back towards wherever it came from.

zshi-redhat commented:
> dont we need an OVN bump to pick up some OVN fix that allows handling for this case?

I built the image from master, the tests showed that icmp frag msg was sent from the gateway_port (it has ip.src = rtoj-port ), I assume the current ovn version contains the fix, but let me double check with Dumitru.

zshi-redhat commented:
> "@trozet When pkt larger than mtu hit check_pkt_larger in the physical bridge br-ex and be forwarded to host kernel, how does it then get to the ovn pod?"
> 
> @zshi-redhat if I understand your question/scenario correctly...the kernel will reply with ICMP too large and send it back into br-ex bridge, then out the physical nic, back towards wherever it came from.

@trozet  I ran the large udp pkt test with openshift (w/o gateway_mtu) where reply udp pkt is larger than ovn mtu (1414), tcpdump on the node physical port showed no icmp msg from client kernel side, I'm wondering why it still works (reply pkt be received on the ovn pod).

Compared with tests where gateway_mtu is applied, I did see the icmp msg from client kernel, but it didn't work (reply pkg not received on the ovn pod). Since udp doesn't have retransmission, I think this is the expected result, isn't it?

zshi-redhat commented:
> > dont we need an OVN bump to pick up some OVN fix that allows handling for this case?
> 
> I built the image from master, the tests showed that icmp frag msg was sent from the gateway_port (it has ip.src = rtoj-port ), I assume the current ovn version contains the fix, but let me double check with Dumitru.

When building the image from master + this PR, I used the Dockerfile from openshift repo: https://github.com/openshift/ovn-kubernetes/blob/master/Dockerfile. Talked with Dumitru, he confirms that this downstream ovn 21.09.0-18.el8fdp includes gateway_mtu support.

adrianchiris commented:
> /cc @adrianchiris @girishmg This fixes the east-west HWOL issue that check_pkt_larger is not offloadable yet.

I still need to give this a deeper look but i think we should still have a config knob to disable packet mtu check for deployments who want North/South traffic offloaded (given no MTU issues in their environment).

using OVN  LRP `options:gateway_mtu` would still program `check_pkt_larger` flow in OVS.
https://github.com/ovn-org/ovn/blob/9242f27f6358582995ac1ad06a1f414940f56e82/northd/ovn-northd.8.xml#L1989

adrianchiris commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -152,11 +140,9 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bo
-							cookie, npw.ofportPhys, flowProtocol, svcPort.NodePort, actions),
-						fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_src=%d, "+
-							"actions=output:%s",
+						fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_dst=%d, actions=%s",
```
nit: to keep things consistent maybe use "actions=output:%s", the `output:` prefix is used in other places in this file when adding OF rule which outputs a packet on a certain port ?

zshi-redhat replied:
Unreverted the `disable mtu` flag change.

adrianchiris commented on go-controller/pkg/ovn/gateway_init.go:
```diff
@@ -60,6 +60,9 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
 	for _, gwLRPIfAddr := range gwLRPIfAddrs {
 		args = append(args, gwLRPIfAddr.String())
 	}
+	maxPktLength := config.Default.MTU + 14
```
nit: define 14 as a meaningful const ?

zshi-redhat commented:
> /retest @zshi-redhat Fedora OVN has been bumped to 21.09 and should include gateway_mtu support now.

Rebased to latest master and fixed comments.

zshi-redhat commented:
/retest

adrianchiris commented on dist/images/ovnkube.sh:
```diff
@@ -867,9 +867,9 @@ ovn-master() {
-  disable_pkt_mtu_check_flag=
-  if [[ ${ovn_disable_pkt_mtu_check} == "true" ]]; then
-      disable_pkt_mtu_check_flag="--disable-pkt-mtu-check"
+  disable_gateway_mtu_flag=
```
This is an ovnkube-master configuration so need to pass it to ovnkube.sh cmd line in line 927 so this conifg is populated in master

adrianchiris commented on dist/images/ovnkube.sh:
```diff
@@ -1140,7 +1140,7 @@ ovn-node() {
     ${hybrid_overlay_flags} \
     ${disable_snat_multiple_gws_flag} \
-    ${disable_pkt_mtu_check_flag} \
+    ${disable_gateway_mtu_flag} \
```
ovnkube-node does not care for this config right ? so can be removed IMO

zshi-redhat replied:
Done

adrianchiris commented on dist/templates/ovnkube-node.yaml.j2:
```diff
@@ -151,8 +151,8 @@ spec:
           value: "{{ ovn_disable_snat_multiple_gws }}"
-        - name: OVN_DISABLE_PKT_MTU_CHECK
-          value: "{{ ovn_disable_pkt_mtu_check }}"
+        - name: OVN_DISABLE_GATEWAY_MTU
```
need to move this to ovnkube-master.yaml.j2 template

zshi-redhat replied:
Moved to ovnkube-master

adrianchiris commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -152,11 +142,9 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bo
-							cookie, npw.ofportPhys, flowProtocol, svcPort.NodePort, actions),
-						fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, tp_src=%d, "+
-							"actions=output:%s",
+						fmt.Sprintf("cookie=%s, priority=100, in_port=%s, %s, tp_dst=%d, actions=%s",
```
nodeport related flows have priority 110, why was this changed to 100 ?

adrianchiris replied:
this PR bumped node port related flows priority to 110 

https://github.com/ovn-org/ovn-kubernetes/commit/2c0ec2337e590e2287c57f80975dadb79a9ad1cf

zshi-redhat replied:
I mistakenly changed this when reverting the commit. fixed.

zshi-redhat commented:
/retest

dcbw commented:
@zshi-redhat could you close this PR, rebase, and submit another one with the rebased branch?

I don't know why the CI keeps canceling, but it may be related to the fact that when you opened this PR you weren't an org member but got added later. A new PR should immediately start CI and let it run.

zshi-redhat commented:
> @zshi-redhat could you close this PR, rebase, and submit another one with the rebased branch?
> 
> I don't know why the CI keeps canceling, but it may be related to the fact that when you opened this PR you weren't an org member but got added later. A new PR should immediately start CI and let it run.

I recreated the PR at https://github.com/ovn-org/ovn-kubernetes/pull/2620

zshi-redhat commented:
superseded by https://github.com/ovn-org/ovn-kubernetes/pull/2620, closing.
